### PR TITLE
[TRACKER] Master Mirror

### PR DESCRIPTION
**Pull request by NoriteSC**
Contains all the changes made to this engine
and will include incoming changes

the Viewport camera and editor viewport rewrite and better gizmo, has ben moved to https://github.com/FlaxEngine/FlaxEngine/pull/1633
# **Avating**
- [ ]  re work the numpad controls to be what blender has
   - numpad  2 rotate down
   - numpad  3 right
   - numpad  4 rotate left
   - numpad  5 toggle ortho mode
   - numpad  6 rotate right
   - numpad  7 up
   - numpad  8 rotate up
   - numpad  9 swap betwine + and - view

- [ ] https://github.com/FlaxEngine/FlaxEngine/issues/1342
- [ ] https://github.com/FlaxEngine/FlaxEngine/issues/1227
- [ ] https://github.com/FlaxEngine/FlaxEngine/issues/1126
- [ ] https://github.com/FlaxEngine/FlaxEngine/issues/1056
- [ ] Actor icons in scene window
       image

- [ ] content browsher [enhancement]
     - [ ] https://github.com/FlaxEngine/FlaxEngine/issues/1132#issue-1731306489
     - [ ] colored folders
     - [ ] Add icons for prefab variants 
       https://github.com/FlaxEngine/FlaxEngine/issues/1452#issue-1903971262
- [ ] Better Profiler
**issues**
     - [x] https://github.com/FlaxEngine/FlaxEngine/issues/1380

- [ ] simplify code and fix the Auto Target (a part of https://github.com/FlaxEngine/FlaxEngine/pull/1337 Joint-related stuff was reverted because is breaking [PhysicsFeaturesTour](https://github.com/FlaxEngine/FlaxSamples/tree/master/PhysicsFeaturesTour)) the discussion https://github.com/FlaxEngine/FlaxEngine/issues/1524

- [ ] Player Actor

- [ ] https://github.com/FlaxEngine/FlaxEngine/issues/1173

- [ ] World composicion
**work needed**
- - [ ] Add sector fields scene transform
- - [ ] Update the EditorViewport to content the sector field
- - [ ] Async load unload on the fly by EditorViewport
- - [ ] World bounds
- **Info**
Sector size 20,000 (float) units 
Max number of sectors in the world 19 807 040 600 895 968 300 706 562 046
Max number of sectors per axis 4 294 967 294 

- [ ] Server meshing (linked servers in local network) for World composicion sectors
# Maybe Avating
runtime simple profiler somfing like ( https://github.com/Tayx94/graphy)

# **Incudes the** 
- https://github.com/FlaxEngine/FlaxEngine/pull/1290
- https://github.com/FlaxEngine/FlaxEngine/pull/1337
- https://github.com/FlaxEngine/FlaxEngine/pull/1376
- https://github.com/FlaxEngine/FlaxEngine/pull/1378
- https://github.com/FlaxEngine/FlaxEngine/pull/1412
- https://github.com/FlaxEngine/FlaxEngine/pull/1377
- https://github.com/FlaxEngine/FlaxEngine/pull/1581
- https://github.com/FlaxEngine/FlaxEngine/pull/1600
- https://github.com/FlaxEngine/FlaxEngine/pull/1606
- https://github.com/FlaxEngine/FlaxEngine/pull/1628
- https://github.com/FlaxEngine/FlaxEngine/pull/1633
# **For manteners**
 cheay pick what looks nice or is needed
 any new stuff from me is going directy to this
 and any pr with pokes in the stuff modified by me, with has a use 
 so keep a look at **Incudes the**  when Reviewering

### CODE REVIEW

**Comment by stefnotch** [comment]
@NoriteSC You can upload the assets as a .zip, together with a note about where we should put them.

**Comment by NoriteSC** [comment]
> @NoriteSC You can upload the assets as a .zip (right here in a GitHub comment), together with a note about where we should put them.

i will add the assets when feature is ready to be picked

File: Source/Editor/Editor.cs [hunk 1]
```diff
@@ -248,6 +247,11 @@ static Editor()
         /// </summary>
         public event Action PlayModeEnd;
 
+        /// <summary>
+        /// Occurs when play mode ends (after leaving play mode).
```
Comment (stefnotch) [inline Source/Editor/Editor.cs:251]: Copy paste error, I suppose?

File: Source/Editor/Gizmo/TransformGizmoBase.Draw.cs [hunk 2]
```diff
@@ -6,206 +6,260 @@ namespace FlaxEditor.Gizmo
...
+                            break;
+                        case Axis.YZ:
+                            DebugDraw.DrawLine(Vector3.Forward * WorldTransform.Orientation * 100000f + WorldTransform.Translation, Vector3.Backward * WorldTransform.Orientation * 100000f + WorldTransform.Translation, Resources.YAxisColor, 0f, true);
+                            DebugDraw.DrawLine(Vector3.Up * WorldTransform.Orientation * 100000f + WorldTransform.Translation, Vector3.Down * WorldTransform.Orientation * 100000f + WorldTransform.Translation, Resources.ZAxisColor, 0f, true);
+                            break;
+                    }
+                    bool flag2 = _activeMode == Mode.Rotate;
+                    if (flag2)
+                    {
+                        bool flag3 = !dir.IsZero;
```
Comment (stefnotch) [inline Source/Editor/Gizmo/TransformGizmoBase.Draw.cs:68]: Aw, that's where the Git thingy, followed by decompiling your own code happened, right?

File: Source/Engine/Core/Math/Vector2.cs [hunk 3]
```diff
@@ -1380,7 +1380,27 @@ public static Vector2 TransformNormal(Vector2 normal, Matrix transform)
...
+        /// </summary>
+        /// <param name="vector">Refrance vector</param>
+        /// <param name="GridSize">Grid cell size</param>
+        /// <returns></returns>
+        public static Vector2 SnapToGrid(Vector2 vector, Real GridSize)
+        {
+            vector -= (GridSize * 0.5f);
+            vector /= GridSize;
+            return Ceil(vector) * GridSize;
+        }
```
Comment (stefnotch) [inline Source/Engine/Core/Math/Vector2.cs:1394]: This seems equivalent to 
```csharp
 public static Vector2 RoundTo(Vector2 vector, Real step)
        {
            return Round(vector / step) * step;
        }
```

Well, except for the very edge cases. `RoundTo(new Vector2(5, 0), 10)` will round up, while SnapToGrid will round down

File: Source/Engine/Scripting/Internal/EngineInternalCalls.cpp [hunk 4]
```diff
@@ -117,7 +118,7 @@ DEFINE_INTERNAL_CALL(void) ProfilerInternal_BeginEvent(MString* nameObj)
 #if COMPILE_WITH_PROFILER
     StringView name;
     MUtils::ToString(nameObj, name);
-    ProfilerCPU::BeginEvent(*name);
+    ManagedEventsCPU.Push(ProfilerCPU::BeginEvent(*name));
```
Comment (stefnotch) [inline Source/Engine/Scripting/Internal/EngineInternalCalls.cpp:121]: Oh, what's this?

File: Source/Editor/Editor.cs (hunk 1 above)
Comment (NoriteSC) [reply Source/Editor/Editor.cs:251]: yes

File: Source/Engine/Platform/Windows/WindowsWindow.cpp [hunk 5]
```diff
@@ -1178,7 +1178,7 @@ LRESULT WindowsWindow::WndProc(UINT msg, WPARAM wParam, LPARAM lParam)
         }
         OnLostFocus();
         break;
-    case WM_ACTIVATEAPP:
+    case WM_ACTIVATE:
```
Comment (stefnotch) [inline Source/Engine/Platform/Windows/WindowsWindow.cpp:1181]: Forgive my absolute noob question, but is switching the message type here the right thing to do?
WM_ACTIVATE seems to have a 3rd possible value `WA_CLICKACTIVE 2 Activated by a mouse click.`, which should probably also be handled?

File: Source/Editor/Gizmo/TransformGizmoBase.Draw.cs (hunk 2 above)
Comment (NoriteSC) [reply Source/Editor/Gizmo/TransformGizmoBase.Draw.cs:68]: yes

File: Source/Engine/Scripting/Internal/EngineInternalCalls.cpp (hunk 4 above)
Comment (NoriteSC) [reply Source/Engine/Scripting/Internal/EngineInternalCalls.cpp:121]: small fix for c# side what is removed and added to the Profiler stack the GPU profiler has this so i added it to CPU side

File: Source/Engine/Platform/Base/WindowBase.cpp [hunk 6]
```diff
@@ -439,6 +440,9 @@ void WindowBase::OnGotFocus()
         return;
     _focused = true;
 
+    if (RenderTask && _settings.WindowFPSWhenNotFocused > ZeroTolerance)
```
Comment (stefnotch) [inline Source/Engine/Platform/Base/WindowBase.cpp:443]: Is `_settings.WindowFPSWhenNotFocused` an integer? If so, just writing `_settings.WindowFPSWhenNotFocused > 0` should do the trick.

Comment (NoriteSC) [reply Source/Engine/Platform/Base/WindowBase.cpp:443]: no my code :D

File: Source/Engine/Core/Math/Vector2.cs (hunk 3 above)
Comment (NoriteSC) [reply Source/Engine/Core/Math/Vector2.cs:1394]: it is close but the SnapToGrid snaps to the center of grid cell and the offset make it consistent where the scap acures

**Comment by NoriteSC** [comment]
hard reset

**Pull request closed without merging**